testsuite/smoke-test/src/smoke_test_environment.rs

```diff
@@ -58,66 +66,116 @@ impl SmokeTestEnvironment {
     }
 
     pub fn setup_vfn_swarm(&mut self) {
-        self.vfn_swarm = Some(
-            DiemSwarm::configure_fn_swarm(
-                &workspace_builder::get_diem_node_with_failpoints(),
-                None,
-                None,
-                &self.validator_swarm.config,
-                FullnodeType::ValidatorFullnode,
-            )
-            .unwrap(),
-        );
+        let swarm_key = FullNodeSwarmType::Validator;
+        if self.fn_swarms.contains_key(&swarm_key) {
+            panic!("Already setup VFN swarm");
+        }
+
+        let swarm = DiemSwarm::configure_fn_swarm(
+            "ValidatorFullNode",
+            &workspace_builder::get_diem_node_with_failpoints(),
+            None,
+            None,
+            &self.validator_swarm.config,
+            FullnodeType::ValidatorFullnode,
+        )
+        .unwrap();
+        self.add_fn_swarm(swarm_key, swarm);
     }
 
-    pub fn setup_pfn_swarm(&mut self, num_nodes: usize) {
-        self.pfn_swarm = Some(
-            DiemSwarm::configure_fn_swarm(
-                &workspace_builder::get_diem_node_with_failpoints(),
-                None,
-                None,
-                &self.validator_swarm.config,
-                FullnodeType::PublicFullnode(num_nodes),
-            )
-            .unwrap(),
-        );
+    pub fn add_public_fn_swarm(
+        &mut self,
+        name: &'static str,
+        num_nodes: usize,
+        template: Option<NodeConfig>,
+        upstream_swarm: &SwarmConfig,
+    ) {
+        // Let's shortcut it so we don't have to wait for any startup time
+        let swarm_key = FullNodeSwarmType::Public(name);
+        if self.fn_swarms.contains_key(&swarm_key) {
+            println!("Already setup {:?} swarm", swarm_key);
+            panic!("Already setup full node swarm");
```

Comment:
Why the two lines? just have the println messge as the panic one.

Reply:
Hmm, don't know probably had to deal with debugging how I mentioned earlier kinda scrappy and confused between swarms.